internal/hcsoci/credentials.go

```diff
@@ -0,0 +1,115 @@
+func (instance *CCGState) Release(ctx context.Context) error {
+	if err := removeCredentialGuard(ctx, instance.id); err != nil {
+		log.G(ctx).WithFields(logrus.Fields{
+			logrus.ErrorKey: err,
+			"containerID":   instance.id,
+		}).Error("failed to remove container credential guard instance")
+		return err
+	}
+	return nil
+}
+
+// CreateCredentialGuard creates a container credential guard instance and returns the state object to be placed in a v2 container doc.
+func CreateCredentialGuard(ctx context.Context, id, credSpec string, hypervisorIsolated bool) (*hcsschema.ContainerCredentialGuardState, error) {
+	log.G(ctx).WithField("containerID", id).Debug("creating container credential guard instance")
+	// V2 schema ccg setup a little different as its expected to be passed
+	// through all the way to the gcs. Can no longer be enabled just through
+	// a single property. The flow is as follows
+	// ------------------------------------------------------------------------
+	// 1. Call HcsModifyServiceSettings with a ModificationRequest set with a
+	// ContainerCredentialGuardAddInstanceRequest. This is where the cred spec
+	// gets passed in. Transport either "LRPC" (Argon) or "HvSocket" (Xenon).
+	// 2. Query the instance with a call to HcsGetServiceProperties with the
+	// PropertyType "ContainerCredentialGuard". This will return all instances
+	// 3. Parse for the id of our container to find which one correlates to the
+	// container we're building the doc for, then add to the V2 doc.
+	// 4. If xenon container the hvsocketconfig will need to be in the UVMs V2 schema
+	// HcsComputeSystem document before being created/sent to HCS. It must be in the doc
+	// at creation time as we do not support hot adding hvsocket service table entries.
+	// This is currently a blocker for adding support for hyper-v gmsa.
+	transport := "LRPC"
+	if hypervisorIsolated {
+		// TODO(Dcantah) Set transport to HvSocket here when this is supported
+		return nil, fmt.Errorf("v2 hyper-v isolated containers do not support gmsa")
+	}
+	req := hcsschema.ModificationRequest{
+		PropertyType: hcsschema.PTContainerCredentialGuard,
+		Settings: &hcsschema.ContainerCredentialGuardOperationRequest{
+			Operation: hcsschema.AddInstance,
+			OperationDetails: &hcsschema.ContainerCredentialGuardAddInstanceRequest{
+				Id:             id,
+				CredentialSpec: credSpec,
+				Transport:      transport,
+			},
+		},
+	}
+	if err := hcs.ModifyServiceSettings(ctx, req); err != nil {
+		return nil, fmt.Errorf("failed to generate container credential guard instance")
+	}
+
+	q := hcsschema.PropertyQuery{
+		PropertyTypes: []hcsschema.PropertyType{hcsschema.PTContainerCredentialGuard},
+	}
+	serviceProps, err := hcs.GetServiceProperties(ctx, q)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve container credential guard instances")
+	}
+	if len(serviceProps.Properties) != 1 {
+		return nil, fmt.Errorf("wrong number of service properties present")
```

Comment:
Prefer `errors.New` if you aren't actually using the string formatting functionality.